### PR TITLE
fix(ci): lock file validation

### PR DESCRIPTION
**Pull request by ruyadorno**
Make sure to validate any lock file (either `package-lock.json` or
`npm-shrinkwrap.json`) against the current install. This will properly
throw an error in case any of the dependencies being installed don't
match the dependencies that are currently listed in the lock file.

## References
Fixes: https://github.com/npm/cli/issues/2701
Fixes: https://github.com/npm/cli/issues/3947

### CODE REVIEW

File: lib/commands/ci.js [hunk 1]
```diff
@@ -55,6 +56,28 @@ class CI extends ArboristWorkspaceCmd {
...
+
+    // verifies that the packages from the ideal tree will match
+    // the same versions that are present in the virtual tree (lock file)
+    // throws a validation error in case of mismatches
+    const errors = validateLockfile(virtualInventory, arb.idealTree.inventory)
+    if (errors.length) {
+      throw new Error(
+        '`npm ci` can only install packages when your package.json and ' +
+        'package-lock.json or npm-shrinkwrap.json are in sync. Please ' +
+        'update your lock file with `npm install` ' +
```
Comment (wraithgar) [inline lib/commands/ci.js:75]: 💯 having a path to resolution here explicitly laid out

Comment (ruyadorno) [reply lib/commands/ci.js:75]: agreed! to be fair it's not my doing 😁 I'm just reinstating the message from v6: 

https://github.com/npm/libcipm/blob/9ab1a620db485c137b1c89979c80beddf7e2da42/index.js#L172-L178

File: lib/commands/ci.js [hunk 2]
```diff
@@ -55,6 +56,28 @@ class CI extends ArboristWorkspaceCmd {
...
+    const virtualInventory = new Map(arb.virtualTree.inventory)
+
+    // build ideal tree step needs to come right after retrieving the virtual
+    // inventory since it's going to erase the previous ref to virtualTree
+    await arb.buildIdealTree()
+
+    // verifies that the packages from the ideal tree will match
+    // the same versions that are present in the virtual tree (lock file)
+    // throws a validation error in case of mismatches
+    const errors = validateLockfile(virtualInventory, arb.idealTree.inventory)
```
Comment (wraithgar) [inline lib/commands/ci.js:70]: This is a single-use function for this command only.  Why extract it into a separate file?  Can we leave it in here, and let the tests be true integration tests for CI?

Comment (ruyadorno) [reply lib/commands/ci.js:70]: it sounds nice but the reality is that the tests we have in place for `npm ci` today are also super unitary tests mocking the world and testing the implementation, we would end up with the same kind of tests but with 10x more pieces floating around / mocked surface.

I would push back for having the refactor of the tests as a whole to be a tackled as part of a different story - if that's what we want.

Comment (wraithgar) [reply lib/commands/ci.js:70]: wfm

**Review by wraithgar — APPROVED** [review]

**Comment by das7pad** [comment]
Hi! Are there plans for back-porting this fix to npm@v7?

**Pull request merged**